File: Jim/src/sk/fiit/jim/agent/highskill/WalkNew.java
package sk.fiit.jim.agent.highskill;

import sk.fiit.jim.agent.AgentInfo;
import sk.fiit.jim.agent.moves.LowSkill;
import sk.fiit.jim.agent.moves.LowSkills;
import sk.fiit.jim.agent.skills.HighSkill;
import sk.fiit.robocup.library.geometry.Angles;
import sk.fiit.robocup.library.geometry.Vector3D;
import sk.fiit.jim.agent.models.AgentModel;
import sk.fiit.jim.agent.models.EnvironmentModel;
import sk.fiit.jim.agent.models.WorldModel;

/**
 * 
 * Reimplemented from Ruby to Java by Stefan Horvath
 *
 */
public class WalkNew extends HighSkill {
	
	AgentInfo agentInfo = AgentInfo.getInstance();
	AgentModel agentModel = AgentModel.getInstance();
	
	Angles right = new Angles(180, 360);
	Angles left = new Angles(0, 180);
	Angles front1 = new Angles(0, 90);
	Angles front2 = new Angles(270, 360);
	Angles back = new Angles(90, 270);
	
	Angles go1 = new Angles(0, 30);
	Angles go2 = new Angles(330, 360);

	Angles range1 = new Angles(0.0, 25.0);
	Angles range2 = new Angles(335.0, 360.0);
	
	boolean haveToBeTurnedToTarget;
	double radiusBall;
	double radiusSpeed;
	
	Vector3D ball = null;
	Vector3D kickTarget = null;
	
	public WalkNew(boolean haveToBeTurnedToTarget, double radiusBall, double radiusSpeed) {
		super();
		this.haveToBeTurnedToTarget = haveToBeTurnedToTarget;
		this.radiusBall = radiusBall;
		this.radiusSpeed = radiusSpeed;
		
		if (haveToBeTurnedToTarget) {
			go1 = new Angles(0, 15);
			go2 = new Angles(335, 360);
		}
		else {
			go1 = new Angles(0, 30);
			go2 = new Angles(330, 360);
		}
	}
	//Todo #Task(Implement validity_proc) #Solver(xmarkech) #Priority() | xmarkech 2013-12-10T20:27:54.6970000Z 
	
	
	@Override
	public LowSkill pickLowSkill() {
		ball = WorldModel.getInstance().getBall().getRelativePosition();
		kickTarget = agentInfo.kickTarget();
		
		if (false) { // ruby --> validity_proc ???
			agentInfo.loguj("validity");
			return null;
		}
		else if (ball.getR() < radiusBall) {
			agentInfo.loguj("ball radius");
			return null;
		}
		else if (EnvironmentModel.beamablePlayMode()) {
			agentInfo.loguj("beam");
			return null;
		}
		else if (agentModel.isOnGround()) {
			agentInfo.loguj("fall");
			return null;
		}
		else if (ballUnseen() > 3) {
			agentInfo.loguj("ball unseen");
			return null;
		}
		else if (haveToBeTurnedToTarget && !turnedToGoal()) {
			agentInfo.loguj("not turned to goal");
			return null;
		}
		else if (ballInRange(go1) || ballInRange(go2)) {
			if (closeEnough(radiusSpeed)) {
				agentInfo.loguj("go and blizko");
				return LowSkills.get("walk_forward");
			}
			else {
				agentInfo.loguj("go and daleko");
				return LowSkills.get("walk_forward");
			}
		}
		else if (ballInRange(left) && (ballInRange(front1) || ballInRange(front2))) {
			agentInfo.loguj("vlavo vpredu");
			if ((ball.getR() < 0.5) && haveToBeTurnedToTarget)
				return LowSkills.get("walk_forward");
			else
				return LowSkills.get("turn_left_cont_20");
		}
		else if (ballInRange(right) && (ballInRange(front1) || ballInRange(front2))) {
			agentInfo.loguj("vpravo vpredu");
			if ((ball.getR() < 0.5) && haveToBeTurnedToTarget)
				return LowSkills.get("walk_forward");
			else
				return LowSkills.get("turn_right_cont_20");
		}
		else if (ballInRange(left) && ballInRange(back)) {
			agentInfo.loguj("vlavo vzadu");
			return LowSkills.get("turn_left_90");
		}
		else if (ballInRange(right) && ballInRange(back)) {
			agentInfo.loguj("vpravo vzadu");
			return LowSkills.get("turn_right_90");
		}
		return null;
	}
	
	public double ballUnseen() {
		return EnvironmentModel.SIMULATION_TIME - WorldModel.getInstance().getBall().getLastTimeSeen();
	}
	
	public boolean turnedToGoal() {
		return range1.include(kickTarget.getPhi()) || range2.include(kickTarget.getPhi());
	}

	public boolean ballInRange(Angles range) {
		return range.include(Math.toDegrees(ball.getPhi()));
	}
	
	public boolean closeEnough(double radius) {
		return ball.getR() < radius;
	}
	
	@Override
	public void checkProgress() throws Exception {
	}

}

File: Old_xml_and_scripts/Jim/high_skills/new-18.5.2013/walkNew.rb
require 'high_skills/ruby_high_skill.rb'
require 'plan/plan.rb'
include_class Java::sk.fiit.robocup.library.geometry.Angles
include_class Java::java.lang.Math
include_class Java::sk.fiit.jim.agent.models.AgentModel
include_class Java::sk.fiit.jim.agent.models.EnvironmentModel
include_class Java::sk.fiit.jim.agent.models.WorldModel

class WalkNew < RubyHighSkill
  
  @@right = 180..360
  @@left = 0..180
  @@front1 = 0..90
  @@front2 = 270..360
  @@back = 90..270
  
  @@go1 = 0..30
  @@go2 = 330..360
  
  @@range1 = Walk.angle_range(0.0, 25.0)
  @@range2 = Walk.angle_range(335.0, 360.0)
  
  
  
  def initialize haveToBeTurnedToTarget,radius_ball,radius_speed,validity_proc
    super()
    @haveToBeTurnedToTarget = haveToBeTurnedToTarget
    @radius_ball = radius_ball
    @radius_speed = radius_speed
    @validity_proc = validity_proc
    @agentInfo = Java::sk.fiit.jim.agent.AgentInfo.getInstance
    @agentModel = Java::sk.fiit.jim.agent.models.AgentModel.getInstance  
    
    if(@haveToBeTurnedToTarget)
      @@go1 = 0..15
      @@go2 = 335..360
    else
      @@go1 = 0..30
      @@go2 = 330..360
    end
    
  end
  
  def pickLowSkill
    @ball = WorldModel.getInstance.getBall.getRelativePosition
    @kickTarget = @agentInfo.kickTarget()
    
    case
      when (not @validity_proc.call) : 
        puts "validity"
        return nil
      when @ball.r < @radius_ball :
        puts "ball radius" 
        return nil
      when EnvironmentModel.beamablePlayMode :
        puts "beam" 
        return nil
      when @agentModel.falled? :
        puts "fall" 
        return nil
      when ball_unseen > 3 :
        puts "ball unseen" 
        return nil
      when (@haveToBeTurnedToTarget and not turned_to_goal?) 
        puts "not turned to goal"
        return nil
      when ((ballInRange? @@go1) or (ballInRange? @@go2)) :
        if close_enough?(@radius_speed)
          puts "go and blizko"
          return get_skill("walk_forward")
        else
          puts "go and daleko"
          return get_skill("walk_forward")
        end
        
    when (ballInRange? @@left and (ballInRange? @@front1 or ballInRange? @@front2)) :
        puts "vlavo vpredu" 
        if(@ball.r<0.5 and @haveToBeTurnedToTarget)
          return get_skill("walk_forward")
        else
          return get_skill("turn_left_cont_20")
        end
      when (ballInRange? @@right and (ballInRange? @@front1 or ballInRange? @@front2)) :
        puts "vpravo vpredu"
        if (@ball.r<0.5 and @haveToBeTurnedToTarget)
          return get_skill("walk_forward")  
        else
          return get_skill("turn_right_cont_20")
        end
      when (ballInRange? @@left and ballInRange? @@back) :
        puts "vlavo vzadu"  
        return get_skill("turn_left_90")
      when (ballInRange? @@right and ballInRange? @@back) :
        puts "vpravo vzadu"  
        return get_skill("turn_right_90")    
    end    
    return nil
    
  end
  
  def ballInRange? range
    range.include? (Math.toDegrees(@ball.phi))
  end
  
  def close_enough? radius
    @ball.r < radius
  end
  
  def turned_to_goal?
    @@range1.include?(@kickTarget.phi) or @@range2.include?(@kickTarget.phi)
  end
  
  def straight? target_position
    Angles.angle_diff(target_position.phi, 0.0) < Math::PI / 6.0 
  end
  
  def left_and_distant? target_position
    @@left_range.include?(Angles.normalize(target_position.phi))
  end
  
  def right_and_distant? target_position
    @@right_range.include?(Angles.normalize(target_position.phi))
  end
  
  def checkProgress
#    raise "Fallen!" if me.on_ground?
  end

  def ball_unseen
    EnvironmentModel.SIMULATION_TIME - WorldModel.getInstance.getBall.getLastTimeSeen()
  end
end

#in order not to defer the startup cost
#WalkNew.new :ball